### PR TITLE
Bael 5296: Added Http Request Header using the Feign Client

### PR DESCRIPTION
**Pull request by saikatcse03**
In this pr, I have added a few additional packages which contain feign client-specific request header implementation.
Also, looks like earlier PR for Cassandra was not merged.

### CODE REVIEW

File: feign/src/main/java/com/baeldung/feign/clients/builder/BookFeignClientBuilder.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import com.baeldung.feign.header.interceptor.AuthRequestInterceptor;
+
+import feign.Feign;
+import feign.gson.GsonDecoder;
+import feign.gson.GsonEncoder;
+
+
+public class BookFeignClientBuilder {
+
+    public <T> T createClient(Class<T> type, String uri) {
```
Comment (ashleyfrieze) [inline feign/src/main/java/com/baeldung/feign/clients/builder/BookFeignClientBuilder.java:12]: These methods should probably be `static`. There's nothing `instance`ey about this code. We don't access the builder via an interface and the methods don't use any fields of the builder class.

Also, I'm surprised by the use of the Gson library here in that Jackson is more popular. However, it's not relevant to the meat of the article.

File: feign/src/main/java/com/baeldung/feign/header/interceptor/AuthRequestInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+import feign.RequestInterceptor;
+import feign.RequestTemplate;
+
+
+public class AuthRequestInterceptor implements RequestInterceptor {
+   
+    @Override
+    public void apply(RequestTemplate template) {
+        template.header("Authorisation", "Bearer " + this.getAuthorisationToken());
```
Comment (ashleyfrieze) [inline feign/src/main/java/com/baeldung/feign/header/interceptor/AuthRequestInterceptor.java:13]: I think the use of a getter for the auth token is misleading here. This approach would never work in practice. In practice, we'd construct the interceptor by either passing it the auth token to use, or by giving it an object that can fetch the current auth token.

The simplest approach is to make the auth token a String field of this class and to pass that token in on construction. That's real world, and a good example for the reader to follow.

File: feign/src/test/java/com/baeldung/feign/clients/header/dynamicheader/BookClientLiveTest.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import org.junit.Test;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+
+/**
+ * Consumes https://github.com/Baeldung/spring-hypermedia-api
+ */
+@Slf4j
```
Comment (ashleyfrieze) [inline feign/src/test/java/com/baeldung/feign/clients/header/dynamicheader/BookClientLiveTest.java:16]: Let's not use Lombok to create loggers for tests.

Tests shouldn't log, and it makes the test code more oblique to add this.

File: feign/src/test/java/com/baeldung/feign/clients/header/dynamicheader/BookClientLiveTest.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    	
+        Book book = new Book(isbn, "Me", "It's me!", null, null);
+        
+        Map<String,Object> headerMap = new HashMap<>();
+    	
+        headerMap.put("metadata-key1", "metadata-value1");
+        headerMap.put("metadata-key2", "metadata-value2");   	
+    	
+        bookClient.create(headerMap, book);
+        log.info("{}", book);
```
Comment (ashleyfrieze) [inline feign/src/test/java/com/baeldung/feign/clients/header/dynamicheader/BookClientLiveTest.java:42]: Tests should, ideally, end in an assertion.

File: feign/src/test/java/com/baeldung/feign/clients/header/dynamicheader/BookClientLiveTest.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+public class BookClientLiveTest {
+	
+    private BookFeignClientBuilder feignClientBuilder;
+
+    private BookClient bookClient;
+
+    @Before
+    public void setup() {
+        feignClientBuilder = new BookFeignClientBuilder();
+        bookClient = feignClientBuilder.createClient(BookClient.class, "http://localhost:8081/api/books");
```
Comment (ashleyfrieze) [inline feign/src/test/java/com/baeldung/feign/clients/header/dynamicheader/BookClientLiveTest.java:26]: Can we make this test run against a wiremock or mockserver, so the reader can easily run it themselves. I think it's a bad thing to expect them to start up another example app.

File: feign/src/test/java/com/baeldung/feign/clients/header/interceptor/BookClientLiveTest.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+import java.util.UUID;
+
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Consumes https://github.com/Baeldung/spring-hypermedia-api
+ */
+@Slf4j
```
Comment (ashleyfrieze) [inline feign/src/test/java/com/baeldung/feign/clients/header/interceptor/BookClientLiveTest.java:19]: Remove SL4j

File: feign/src/test/java/com/baeldung/feign/clients/header/interceptor/BookClientLiveTest.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    public void givenBookClient_shouldPostBook() throws Exception {
+        String isbn = UUID.randomUUID()
+          .toString();
+        Book book = new Book(isbn, "Me", "It's me!", null, null);
+        bookClient.create(book);
+
+        book = bookClient.findByIsbn(isbn)
+          .getBook();
+        assertThat(book.getAuthor(), is("Me"));
+        log.info("{}", book);
```
Comment (ashleyfrieze) [inline feign/src/test/java/com/baeldung/feign/clients/header/interceptor/BookClientLiveTest.java:59]: remove logging statements

File: feign/src/test/java/com/baeldung/feign/clients/header/staticheader/BookClientLiveTest.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+import java.util.UUID;
+
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Consumes https://github.com/Baeldung/spring-hypermedia-api
+ */
+@Slf4j
```
Comment (ashleyfrieze) [inline feign/src/test/java/com/baeldung/feign/clients/header/staticheader/BookClientLiveTest.java:18]: Remove

File: feign/src/test/java/com/baeldung/feign/clients/header/staticheader/parameterized/BookClientLiveTest.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Consumes https://github.com/Baeldung/spring-hypermedia-api
+ */
+@Slf4j
+public class BookClientLiveTest {
+	
+	private BookFeignClientBuilder feignClientBuilder;
```
Comment (ashleyfrieze) [inline feign/src/test/java/com/baeldung/feign/clients/header/staticheader/parameterized/BookClientLiveTest.java:18]: Indentation

File: feign/src/test/java/com/baeldung/feign/clients/header/staticheader/parameterized/BookClientLiveTest.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Consumes https://github.com/Baeldung/spring-hypermedia-api
+ */
+@Slf4j
```
Comment (ashleyfrieze) [inline feign/src/test/java/com/baeldung/feign/clients/header/staticheader/parameterized/BookClientLiveTest.java:15]: Remove

File: feign/src/test/java/com/baeldung/feign/clients/header/staticheader/parameterized/BookClientLiveTest.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+        feignClientBuilder = new BookFeignClientBuilder();
+        bookClient = feignClientBuilder.createClient(BookClient.class, "http://localhost:8081/api/books");
+    }
+
+    @Test
+    public void givenBookClient_shouldFindOneBook() throws Exception {
+        Book book = bookClient.findByIsbn(requester, "0151072558")
+          .getBook();
+        assertThat(book.getAuthor(), containsString("Orwell"));
+        log.info("{}", book);
```
Comment (ashleyfrieze) [inline feign/src/test/java/com/baeldung/feign/clients/header/staticheader/parameterized/BookClientLiveTest.java:35]: Remove logging statements

File: persistence-modules/java-cassandra/src/main/java/com/baeldung/cassandra/batch/Application.java [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.baeldung.cassandra.batch.domain.Product;
+import com.baeldung.cassandra.batch.repository.KeyspaceRepository;
+import com.baeldung.cassandra.batch.repository.ProductRepository;
+import com.datastax.oss.driver.api.core.CqlSession;
+
+public class Application {
```
Comment (ashleyfrieze) [inline persistence-modules/java-cassandra/src/main/java/com/baeldung/cassandra/batch/Application.java:13]: The cassandra example files seem to be unrelated to this article.

**Review by ashleyfrieze — CHANGES_REQUESTED** [review]

File: feign/src/main/java/com/baeldung/feign/clients/builder/BookFeignClientBuilder.java (hunk 1 above)
Comment (saikatcse03) [reply feign/src/main/java/com/baeldung/feign/clients/builder/BookFeignClientBuilder.java]: Yes. I have changed to static now.  for Gson, i have reused this module so Gson was already used in this module.  But yes Jackson is what i have alos used in all other instances .

File: feign/src/main/java/com/baeldung/feign/header/interceptor/AuthRequestInterceptor.java (hunk 2 above)
Comment (saikatcse03) [reply feign/src/main/java/com/baeldung/feign/header/interceptor/AuthRequestInterceptor.java]: I thought not to make the code for AuthInterceptor too complicated with additional classes, so put one simple method in there. 
Now, I have passing an Interface to this Auth Interceptor and in runtime , i will pass one object which has simple Auth token implementation. So in nutshell, the AuthInterceptor is not generating the token rather delegating to another object instance. Definitely in actual scenario the Overriden class needs to deal with that logic in different way.

File: feign/src/test/java/com/baeldung/feign/clients/header/dynamicheader/BookClientLiveTest.java (hunk 3 above)
Comment (saikatcse03) [reply feign/src/test/java/com/baeldung/feign/clients/header/dynamicheader/BookClientLiveTest.java]: Removed SLF4J from these unit tests.

File: feign/src/test/java/com/baeldung/feign/clients/header/dynamicheader/BookClientLiveTest.java (hunk 4 above)
Comment (saikatcse03) [reply feign/src/test/java/com/baeldung/feign/clients/header/dynamicheader/BookClientLiveTest.java]: Asserts are there. have removed the logs messages

File: feign/src/test/java/com/baeldung/feign/clients/header/dynamicheader/BookClientLiveTest.java (hunk 5 above)
Comment (saikatcse03) [reply feign/src/test/java/com/baeldung/feign/clients/header/dynamicheader/BookClientLiveTest.java]: i feel the wiremock or mockserver will dilute the example for feign and we might need to explain the wiremock itself. So this code is running with actual endpoint which pretty easy to setup.  Also, previous article has use the same example. So, reusability was also a factor for this article.
But definitely, this is very good point to keep in mind.

File: feign/src/test/java/com/baeldung/feign/clients/header/staticheader/parameterized/BookClientLiveTest.java (hunk 9 above)
Comment (saikatcse03) [reply feign/src/test/java/com/baeldung/feign/clients/header/staticheader/parameterized/BookClientLiveTest.java]: corrected

File: feign/src/test/java/com/baeldung/feign/clients/header/staticheader/parameterized/BookClientLiveTest.java (hunk 10 above)
Comment (saikatcse03) [reply feign/src/test/java/com/baeldung/feign/clients/header/staticheader/parameterized/BookClientLiveTest.java]: Removed

File: feign/src/test/java/com/baeldung/feign/clients/header/staticheader/parameterized/BookClientLiveTest.java (hunk 11 above)
Comment (saikatcse03) [reply feign/src/test/java/com/baeldung/feign/clients/header/staticheader/parameterized/BookClientLiveTest.java]: Logging are removed

File: persistence-modules/java-cassandra/src/main/java/com/baeldung/cassandra/batch/Application.java (hunk 12 above)
Comment (saikatcse03) [reply persistence-modules/java-cassandra/src/main/java/com/baeldung/cassandra/batch/Application.java]: I was able to remove this. though these were all merged but still showing in my current branch. but after fetching from upstream master had fixed that.
